include/eosio/vm/signals.hpp

```diff
@@ -27,46 +33,55 @@ namespace eosio { namespace vm {
    template<int Sig>
    inline struct sigaction prev_signal_handler;
 
-   inline bool in_protected_range(void* addr) {
-      //empty protection list means legacy catch-all behavior; useful for some of the old tests
-      if(protected_memory_ranges.empty())
-         return true;
-
-      for(const std::span<std::byte>& range : protected_memory_ranges) {
-         if(addr >= range.data() && addr < range.data() + range.size())
-            return true;
-      }
-      return false;
-   }
-
    inline void signal_handler(int sig, siginfo_t* info, void* uap) {
       sigjmp_buf* dest = std::atomic_load(&signal_dest);
 
-      if (dest && in_protected_range(info->si_addr)) {
-         siglongjmp(*dest, sig);
-      } else {
-         struct sigaction* prev_action;
-         switch(sig) {
-            case SIGSEGV: prev_action = &prev_signal_handler<SIGSEGV>; break;
-            case SIGBUS: prev_action = &prev_signal_handler<SIGBUS>; break;
-            case SIGFPE: prev_action = &prev_signal_handler<SIGFPE>; break;
-            default: std::abort();
+      if (dest) {
+         const void* addr = info->si_addr;
+
+         //neither range set means legacy catch-all behavior; useful for some of the old tests
+         if (code_memory_range.empty() && memory_range.empty())
+            siglongjmp(*dest, sig);
+
+         //a failure in the memory range is always jumped out of
+         if (addr >= memory_range.data() && addr < memory_range.data() + memory_range.size())
+            siglongjmp(*dest, sig);
+
+         //a failure in the code range...
+         if (addr >= code_memory_range.data() && addr < code_memory_range.data() + code_memory_range.size()) {
+            //a SEGV in the code range when timed_run_is_timed_out=false is due to a _different_ thread's execution activating a deadline
+            // timer. Return and retry executing the same code again. Eventually timed_run() on the other thread will reset the page
+            // permissions and progress on this thread can continue
+            if (sig == SIGSEGV && timed_run_is_timed_out.load(std::memory_order_acquire) == false)
```

Comment:
I am suspicious this needs to be `SIGBUS` on macOS. But so far I don't have a good environment to test it with. I think we may need to reevaluate it in the future.